yksmp/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+            // Padding
+            self.align_8();
+            self.read_u16();
+            let num_liveouts = self.read_u16();
+            let liveouts = self.read_liveouts(num_liveouts);
+            self.align_8();
+            v.push(Record {
+                offset,
+                locs,
+                liveouts,
+            });
+        }
+        v
+    }
+
+    fn read_locations(&mut self, num: u16, consts: &Vec<u64>) -> Vec<Location> {
+        let mut v = Vec::new();
+        for _ in 0..num {
+            let kind = self.read_u8();
+            self.read_u8();
+            let _size = self.read_u16();
+            let dwreg = self.read_u16();
+            self.read_u16();
+
+            let location = match kind {
+                0x01 => {
+                    self.read_i32();
+                    Location::Register(dwreg)
+                }
+                0x02 => {
+                    let offset = self.read_i32();
+                    Location::Direct(dwreg, offset)
+                }
+                0x03 => {
+                    let offset = self.read_i32();
+                    Location::Indirect(dwreg, offset)
+                }
+                0x04 => {
+                    let offset = self.read_u32();
+                    Location::Constant(offset)
+                }
+                0x05 => {
+                    let offset = self.read_i32();
+                    Location::LargeConstant(consts[usize::try_from(offset).unwrap()])
+                }
+                _ => unreachable!(),
+            };
+
+            v.push(location)
+        }
+        v
+    }
+
+    fn read_liveouts(&mut self, num: u16) -> Vec<LiveOut> {
+        let mut v = Vec::new();
+        for _ in 0..num {
+            let dwreg = self.read_u16();
+            let size = self.read_u8();
+            v.push(LiveOut { dwreg, size });
+        }
+        v
+    }
+
+    fn align_8(&mut self) {
+        self.offset += (8 - (self.offset % 8)) % 8;
+    }
+
+    fn read_u8(&mut self) -> u8 {
+        let d = u8::from_ne_bytes(self.data[self.offset..self.offset + 1].try_into().unwrap());
+        self.offset += 1;
+        d
+    }
+
+    fn read_u16(&mut self) -> u16 {
+        let d = u16::from_ne_bytes(self.data[self.offset..self.offset + 2].try_into().unwrap());
+        self.offset += 2;
+        d
+    }
+
+    fn read_u32(&mut self) -> u32 {
+        let d = u32::from_ne_bytes(self.data[self.offset..self.offset + 4].try_into().unwrap());
+        self.offset += 4;
+        d
+    }
+
+    fn read_i32(&mut self) -> i32 {
+        let d = i32::from_ne_bytes(self.data[self.offset..self.offset + 4].try_into().unwrap());
+        self.offset += 4;
+        d
+    }
+
+    fn read_u64(&mut self) -> u64 {
+        let d = u64::from_ne_bytes(self.data[self.offset..self.offset + 8].try_into().unwrap());
+        self.offset += 8;
+        d
+    }
+}
+
+/// Maps stackmap records to the addresses where they are relevant. Allows users to retrieve the
```

Comment:
Are they addresses or file offsets?

Reply:
Addresses. We'll later pass `__builtin_return_address(0)` into this to retrieve the correct stackmap record.

Reply:
OK, so should the map be keyed by `usize` then?

Reply:
Since this is all super x64 specific it could also be `u64`.

Reply:
true. just `usize` has a pointer implication, sort of. Up to you.